nx/lib/nx/defn/grad.ex

```diff
@@ -446,6 +446,76 @@ defmodule Nx.Defn.Grad do
     [{x, reduce_g(x, opts, g)}]
   end
 
+  defp grad(:product, [x, opts], ans, g) do
+    # From https://people.maths.ox.ac.uk/gilesm/files/NA-08-01.pdf
+    # 2.2.4 we take that the reverse-mode gradient of a determinant
+    # is A_bar = C_bar . C . inv(A^T)
+
+    # We also know that the determinant of a diagonal matrix is the
+    # product of the diagonal items. This means that we can imagine
+    # our condensed product axes as being the diagonal of a matrix
+    # and apply the determinant grad on it.
+
+    # Also, if A is diagonal, A^T = A and inv(A) is as trivial
+    # as inverting the items. Also, C_bar and C are scalars.
+    # This means that a_bar = c_bar |> Nx.multiply(c) |> Nx.divide(a).
+
+    # First, let's condition our input matrix by substituting 0s with 1s
+    # Then, we move the reducing axes into a condensed innermost axis.
+
+    # After we have our matrix well conditioned, we can operate on the
+    # innermost axis as described above, and then we transpose the items
+    # back into place
+
+    axes =
+      case opts[:axes] do
+        [] -> Nx.axes(x.shape)
+        nil -> Nx.axes(x.shape)
+        axes -> axes
+      end
+
+    outermost_axes = Nx.axes(x.shape) -- axes
+    permutation = outermost_axes ++ axes
+
+    inverse_permutation =
+      permutation
+      |> Enum.with_index()
+      |> Enum.sort_by(fn {x, _} -> x end)
+      |> Enum.map(fn {_, i} -> i end)
+
+    zero_selector = Nx.equal(x, 0)
+
+    reduce_axes_size = axes |> Enum.reduce(1, fn axis, acc -> acc * elem(x.shape, axis) end)
+
+    reduced_shape =
+      outermost_axes
+      |> Enum.map(&elem(x.shape, &1))
+      |> List.insert_at(-1, reduce_axes_size)
+      |> List.to_tuple()
+
+    formatted_x =
+      zero_selector
```

Comment:
The way that Jax defines the result for 0-entries is closer to the forward-mode calculation:

A = product([a b c ... z])
dA = da.b.c....z + a.db.c..... + a.b...dz
and we can notice that each component of the addition on the right side is the same as:

A*da/a, A*db/b and so on (which is closer to the definition I reached as well).
However, this last definition is ill-defined when a given entry is 0.

@jonatanklosko any insights given this?

Reply:
For now the code doesn't break, but also doesn't reach the same result as Jax for 0 entries

Reply:
Does jax special case or do they use the more explicit implementation?

Reply:
They use a different implementation in which ijstead of taking the product and dividing by each item, they take the product with each item removed.

I didn't mimic the implementation at first because they define only the forward-mode derivative. But looking at the current state, I'm thinking about using while to calculate this in a more iterative way.

WDYT?

Reply:
FTR we discussed offline and have some ideas :)